lte/gateway/c/li_agent/PDUGenerator.cpp

```diff
@@ -0,0 +1,166 @@
+/**
+ * Copyright 2020 The Magma Authors.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <cassert>
+#include <iostream>
+#include <string>
+
+#include <netinet/in.h>
+#include <netinet/ip.h>
+#include <net/if.h>
+#include <net/ethernet.h>
+#include <arpa/inet.h>
+
+#include <tins/tins.h>
+#include <tins/ip.h>
+
+#include "PDUGenerator.h"
+
+const PDU::PDUType LIX3PDU::pdu_flag = PDU::USER_DEFINED_PDU;
+
+/* TODO remove after testing */
+void print_bytes(void* ptr, int size) {
+  unsigned char* p = (unsigned char*) ptr;
+  int i;
+  for (i = 0; i < size; i++) {
+    printf("%02hhX ", p[i]);
+  }
+  printf("\n");
+}
+
+void set_xid(struct pdu_info* pdu, std::string value) {
+  memcpy(pdu->xid, value.c_str(), XID_LENGTH);
+}
+
+bool extract_ip_addr(
+    const u_char* packet, std::string& src_ip, std::string& dst_ip) {
```

Comment:
I can wrap it in a class, are you just thinking of something like this? I want to avoid copying data so I don't want to add any append method logic to the class. I could add a requested_length param to the get function so that we can maintain  how much data was actually written/error if the requested_length is more than whats left?

```
class LIPacket {
 public:
  LIPacket(size_t len);

 // ret pointer for writing with specified offset
 get_data(size_t offset);

 private:
  void* data;
  size_t pkt_len;
};
```